File: py110/small_problems_110_119/medium1/9.py
# https://launchschool.com/exercises/030ec9a5

memo = {}

def fibonacci(n):
    if n <= 2:
        return 1
    
    if n not in memo:
        memo[n] = fibonacci(n - 1) + fibonacci(n - 2)
    
    return memo[n]

print(fibonacci(1) == 1)         # True
print(fibonacci(2) == 1)         # True
print(fibonacci(3) == 2)         # True
print(fibonacci(4) == 3)         # True
print(fibonacci(5) == 5)         # True
print(fibonacci(6) == 8)         # True
print(fibonacci(12) == 144)      # True
print(fibonacci(20) == 6765)     # True

File: ls215/smallProblems/med1/9.js
// https://launchschool.com/exercises/9b1769ca


'use strict';

let memo = {}

function fibonacci(n) {
  if (n <= 2) return 1;
  if (n in memo) return memo[n];
  memo[n] = fibonacci(n - 1) + fibonacci(n - 2);
  return memo[n];
}







console.log(fibonacci(1));       // 1
console.log(fibonacci(2));       // 1
console.log(fibonacci(3));       // 2
console.log(fibonacci(4));       // 3
console.log(fibonacci(5));       // 5
console.log(fibonacci(12));      // 144
console.log(fibonacci(1000));      // 6765